In [1]:
import numpy as np
import pandas as pd

In [2]:
import os
os.listdir()

['Mumbai',
 'Bangalore',
 'Hyderabad.csv',
 'Delhi.csv',
 '.DS_Store',
 'Chennai.csv',
 'Mumbai.csv',
 'Data_Cleaning.ipynb',
 'Hyderabad',
 'Chennai',
 'Final_Dataset.csv',
 'Delhi',
 'Data_Understanding.ipynb',
 'Bangalore.csv',
 'Final_Dataset.ipynb',
 '.ipynb_checkpoints']

In [3]:
df=pd.read_csv('Final_Dataset.csv')

In [4]:
df.describe()

Name Original_price    Offer  \
count                                   14180          14180    14180   
unique                                   2195           1460       39   
top     Capital O 2083 Hotel Blessings Deluxe          ₹1777  32% off   
freq                                       30            675    11272   

       Final_price Distance Rating Number of Rating    Summary    Location  \
count        14180    14180  14180            12901      12901       14180   
unique        1231      199     43             1002          5         100   
top          ₹1010   4.4 km    NEW    (164 Ratings)  Very Good  Mahipalpur   
freq          1082      202   1279               83       5999         340   

         City  
count   14180  
unique      5  
top     Delhi  
freq     4280

In [5]:
df.describe(include='all')

Name Original_price    Offer  \
count                                   14180          14180    14180   
unique                                   2195           1460       39   
top     Capital O 2083 Hotel Blessings Deluxe          ₹1777  32% off   
freq                                       30            675    11272   

       Final_price Distance Rating Number of Rating    Summary    Location  \
count        14180    14180  14180            12901      12901       14180   
unique        1231      199     43             1002          5         100   
top          ₹1010   4.4 km    NEW    (164 Ratings)  Very Good  Mahipalpur   
freq          1082      202   1279               83       5999         340   

         City  
count   14180  
unique      5  
top     Delhi  
freq     4280

In [6]:
df['Rating'].value_counts()

NEW     1279
4.2     1057
4.1     1038
4.3      942
4.0      878
3.9      803
4.4      734
3.8      672
4.5      655
3.6      594
3.7      516
4.6      463
4.4      399
3.5      378
4.3      277
3.3      273
4.5      267
4.2      249
4        234
3.4      219
4.7      204
4.8      179
4.1      179
3.9      175
3.2      174
4.6      164
3.6      146
3.8      135
3.1      133
3.7      121
3.0      113
4.7      100
3.4       73
3.2       69
4.8       57
5.0       55
3.3       49
3.1       44
3.5       29
4.9       26
4.0       12
4.9        8
3          8
Name: Rating, dtype: int64

In [7]:
df['Number of Rating'].isnull().sum()

1279

In [8]:
df['Summary'].isnull().sum()

1279

- We can see that there are 14180 hotels in the list.
- We can also understand that number of ratings and summary is missing for every 'NEW' Hotel.

In [9]:
df.head()

Name Original_price    Offer Final_price  \
0       OYO 22085 Raj Banjara White House          ₹2459  32% off       ₹1677   
1         Collection O 42514 Venkat Nagar          ₹3158  32% off       ₹2153   
2  Capital O 44354 Jubilee Banjara Suites          ₹1778  32% off       ₹1213   
3             OYO 35487 Hotel Meena Grand          ₹2090  32% off       ₹1425   
4               OYO 16018 Rainbow Saphire          ₹2607  32% off       ₹1777   

  Distance Rating Number of Rating    Summary       Location       City  
0   1.7 km    4.4    (201 Ratings)  Very Good  Banjara Hills  Hyderabad  
1   1.6 km    4.4    (879 Ratings)  Very Good  Banjara Hills  Hyderabad  
2   1.1 km    4.4   (1258 Ratings)  Very Good  Banjara Hills  Hyderabad  
3   4.0 km      4    (575 Ratings)  Very Good  Banjara Hills  Hyderabad  
4   4.6 km    4.7    (978 Ratings)  Excellent  Banjara Hills  Hyderabad

In [10]:
df.shape

(14180, 10)

In [11]:
df.columns

Index(['Name', 'Original_price', 'Offer', 'Final_price', 'Distance', 'Rating',
       'Number of Rating', 'Summary', 'Location', 'City'],
      dtype='object')

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14180 entries, 0 to 14179
Data columns (total 10 columns):
Name                14180 non-null object
Original_price      14180 non-null object
Offer               14180 non-null object
Final_price         14180 non-null object
Distance            14180 non-null object
Rating              14180 non-null object
Number of Rating    12901 non-null object
Summary             12901 non-null object
Location            14180 non-null object
City                14180 non-null object
dtypes: object(10)
memory usage: 1.1+ MB


In [13]:
df.dtypes

Name                object
Original_price      object
Offer               object
Final_price         object
Distance            object
Rating              object
Number of Rating    object
Summary             object
Location            object
City                object
dtype: object

In [14]:
df.isnull().sum()

Name                   0
Original_price         0
Offer                  0
Final_price            0
Distance               0
Rating                 0
Number of Rating    1279
Summary             1279
Location               0
City                   0
dtype: int64

In [15]:
df.isnull().any(axis=1)

0        False
1        False
2        False
3        False
4        False
5        False
6        False
7        False
8        False
9        False
10       False
11       False
12        True
13       False
14       False
15       False
16       False
17       False
18       False
19       False
20       False
21       False
22       False
23       False
24       False
25       False
26       False
27       False
28       False
29       False
         ...  
14150    False
14151     True
14152    False
14153    False
14154    False
14155    False
14156    False
14157    False
14158    False
14159     True
14160    False
14161    False
14162     True
14163    False
14164    False
14165     True
14166     True
14167    False
14168    False
14169    False
14170    False
14171    False
14172     True
14173    False
14174    False
14175    False
14176    False
14177    False
14178    False
14179    False
Length: 14180, dtype: bool

In [16]:
df.isnull().any(axis=1).sum()

1279

In [17]:
df['Number of Rating'].fillna('(0 Ratings)',inplace=True)

- We have replaced evey null value in the number of ratings by 0, to maintain the authenticity in the data.

In [18]:
df['Summary'].fillna('Yet to be summarised',inplace=True)

In [19]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14180 entries, 0 to 14179
Data columns (total 10 columns):
Name                14180 non-null object
Original_price      14180 non-null object
Offer               14180 non-null object
Final_price         14180 non-null object
Distance            14180 non-null object
Rating              14180 non-null object
Number of Rating    14180 non-null object
Summary             14180 non-null object
Location            14180 non-null object
City                14180 non-null object
dtypes: object(10)
memory usage: 1.1+ MB


In [20]:
df.isnull().sum()

Name                0
Original_price      0
Offer               0
Final_price         0
Distance            0
Rating              0
Number of Rating    0
Summary             0
Location            0
City                0
dtype: int64

In [21]:
df.isnull().any(axis=1).sum()

0

In [22]:
df['Number of Rating'].value_counts()

(0 Ratings)       1279
(164 Ratings)       83
(279 Ratings)       74
(1 Ratings)         71
(283 Ratings)       65
(114 Ratings)       63
(108 Ratings)       61
(21 Ratings)        56
(33 Ratings)        55
(271 Ratings)       51
(143 Ratings)       50
(161 Ratings)       50
(221 Ratings)       50
(559 Ratings)       49
(79 Ratings)        49
(168 Ratings)       47
(252 Ratings)       47
(152 Ratings)       46
(340 Ratings)       45
(306 Ratings)       44
(294 Ratings)       44
(286 Ratings)       44
(159 Ratings)       44
(81 Ratings)        44
(116 Ratings)       43
(121 Ratings)       43
(266 Ratings)       43
(349 Ratings)       42
(228 Ratings)       42
(134 Ratings)       42
                  ... 
(5926 Ratings)       1
(5773 Ratings)       1
(1425 Ratings)       1
(1252 Ratings)       1
(1973 Ratings)       1
(2805 Ratings)       1
(1893 Ratings)       1
(1848 Ratings)       1
(1535 Ratings)       1
(85 Ratings)         1
(2577 Ratings)       1
(565 Ratings)        1
(491 Rating

In [23]:
df['Summary'].value_counts()

Very Good               5999
Good                    3569
Excellent               2089
Yet to be summarised    1279
Fair                    1155
Fabulous                  89
Name: Summary, dtype: int64

In [24]:
df.describe()

Name Original_price    Offer  \
count                                   14180          14180    14180   
unique                                   2195           1460       39   
top     Capital O 2083 Hotel Blessings Deluxe          ₹1777  32% off   
freq                                       30            675    11272   

       Final_price Distance Rating Number of Rating    Summary    Location  \
count        14180    14180  14180            14180      14180       14180   
unique        1231      199     43             1003          6         100   
top          ₹1010   4.4 km    NEW      (0 Ratings)  Very Good  Mahipalpur   
freq          1082      202   1279             1279       5999         340   

         City  
count   14180  
unique      5  
top     Delhi  
freq     4280

- We can see that '₹' is causing a problem in the Original_price and Final_price to convert the data type from object to float
- Also '% off' will not let us convert the Offer from object to float.  
- 'km' in the Distance will not let us convert Distance into float
- 'NEW' in Rating will not lets us convert the data from object to float for better analyses.
- 'ratings' in Number of Rating will not let us convert the data from object to float.


In [25]:
# Removing '₹'from Original_price and Final_price
df['Original_price']=df['Original_price'].str[1:]
df['Final_price']=df['Final_price'].str[1:]

In [26]:
df.head(5)

Name Original_price    Offer Final_price  \
0       OYO 22085 Raj Banjara White House           2459  32% off        1677   
1         Collection O 42514 Venkat Nagar           3158  32% off        2153   
2  Capital O 44354 Jubilee Banjara Suites           1778  32% off        1213   
3             OYO 35487 Hotel Meena Grand           2090  32% off        1425   
4               OYO 16018 Rainbow Saphire           2607  32% off        1777   

  Distance Rating Number of Rating    Summary       Location       City  
0   1.7 km    4.4    (201 Ratings)  Very Good  Banjara Hills  Hyderabad  
1   1.6 km    4.4    (879 Ratings)  Very Good  Banjara Hills  Hyderabad  
2   1.1 km    4.4   (1258 Ratings)  Very Good  Banjara Hills  Hyderabad  
3   4.0 km      4    (575 Ratings)  Very Good  Banjara Hills  Hyderabad  
4   4.6 km    4.7    (978 Ratings)  Excellent  Banjara Hills  Hyderabad

In [27]:
# Removing '% off' from offer
df['Offer']=df['Offer'].str[:2]

In [28]:
df.head(5)

Name Original_price Offer Final_price  \
0       OYO 22085 Raj Banjara White House           2459    32        1677   
1         Collection O 42514 Venkat Nagar           3158    32        2153   
2  Capital O 44354 Jubilee Banjara Suites           1778    32        1213   
3             OYO 35487 Hotel Meena Grand           2090    32        1425   
4               OYO 16018 Rainbow Saphire           2607    32        1777   

  Distance Rating Number of Rating    Summary       Location       City  
0   1.7 km    4.4    (201 Ratings)  Very Good  Banjara Hills  Hyderabad  
1   1.6 km    4.4    (879 Ratings)  Very Good  Banjara Hills  Hyderabad  
2   1.1 km    4.4   (1258 Ratings)  Very Good  Banjara Hills  Hyderabad  
3   4.0 km      4    (575 Ratings)  Very Good  Banjara Hills  Hyderabad  
4   4.6 km    4.7    (978 Ratings)  Excellent  Banjara Hills  Hyderabad

In [29]:
#Removing 'km' from Distance
df['Distance']=df['Distance'].str[:3]

In [30]:
df.head(5)

Name Original_price Offer Final_price  \
0       OYO 22085 Raj Banjara White House           2459    32        1677   
1         Collection O 42514 Venkat Nagar           3158    32        2153   
2  Capital O 44354 Jubilee Banjara Suites           1778    32        1213   
3             OYO 35487 Hotel Meena Grand           2090    32        1425   
4               OYO 16018 Rainbow Saphire           2607    32        1777   

  Distance Rating Number of Rating    Summary       Location       City  
0      1.7    4.4    (201 Ratings)  Very Good  Banjara Hills  Hyderabad  
1      1.6    4.4    (879 Ratings)  Very Good  Banjara Hills  Hyderabad  
2      1.1    4.4   (1258 Ratings)  Very Good  Banjara Hills  Hyderabad  
3      4.0      4    (575 Ratings)  Very Good  Banjara Hills  Hyderabad  
4      4.6    4.7    (978 Ratings)  Excellent  Banjara Hills  Hyderabad

In [31]:
#Replacing 'NEW' with 0 for better analyses
df['Rating'].replace('NEW','0.0',inplace=True)

In [32]:
df.head(15)

Name Original_price Offer  \
0           OYO 22085 Raj Banjara White House           2459    32   
1             Collection O 42514 Venkat Nagar           3158    32   
2      Capital O 44354 Jubilee Banjara Suites           1778    32   
3                 OYO 35487 Hotel Meena Grand           2090    32   
4                   OYO 16018 Rainbow Saphire           2607    32   
5         Collection O 46198 New Sree Krishna           3964    32   
6                 OYO 14224 Happy Guest house           2373    32   
7               OYO 62547 Dream Connect Stays           1941    32   
8                       OYO 11376 Golden Park           2205    45   
9                       OYO 11316 Athithi Inn           2178    38   
10                       OYO 10272 Hotel Ruma           1777    32   
11                         OYO 18725 Sr Grand           1910    32   
12                    OYO 18302 Pride Banjara           2260    32   
13              OYO 983 Hotel Surya Residency           2811    32   
14  Capital O 46832 Gaayatri Residency Deluxe           2098    32   

   Final_price Distance Rating Number of Rating               Summary  \
0         1677      1.7    4.4    (201 Ratings)             Very Good   
1         2153      1.6    4.4    (879 Ratings)             Very Good   
2         1213      1.1    4.4   (1258 Ratings)             Very Good   
3         1425      4.0      4    (575 Ratings)             Very Good   
4         1777      4.6    4.7    (978 Ratings)             Excellent   
5         2703      6.1    3.9    (989 Ratings)                  Good   
6         1618      1.8    4.6    (448 Ratings)             Excellent   
7         1324      0.6    4.3    (267 Ratings)             Very Good   
8         1212      5.3    4.4   (1570 Ratings)             Very Good   
9         1348      4.8    3.3    (697 Ratings)                  Fair   
10        1212      4.2    4.2    (488 Ratings)             Very Good   
11        1303      3.3    4.4    (294 Ratings)             Very Good   
12        1541      1.1    0.0      (0 Ratings)  Yet to be summarised   
13        1917      3.2    4.2    (264 Ratings)             Very Good   
14        1431      4.1    4.8    (644 Ratings)             Excellent   

         Location       City  
0   Banjara Hills  Hyderabad  
1   Banjara Hills  Hyderabad  
2   Banjara Hills  Hyderabad  
3   Banjara Hills  Hyderabad  
4   Banjara Hills  Hyderabad  
5   Banjara Hills  Hyderabad  
6   Banjara Hills  Hyderabad  
7   Banjara Hills  Hyderabad  
8   Banjara Hills  Hyderabad  
9   Banjara Hills  Hyderabad  
10  Banjara Hills  Hyderabad  
11  Banjara Hills  Hyderabad  
12  Banjara Hills  Hyderabad  
13  Banjara Hills  Hyderabad  
14  Banjara Hills  Hyderabad

In [33]:
#Removing () and ratings from Number of Rating
df['Number of Rating']=df['Number of Rating'].str[1:]

In [34]:
df.head(5)

Name Original_price Offer Final_price  \
0       OYO 22085 Raj Banjara White House           2459    32        1677   
1         Collection O 42514 Venkat Nagar           3158    32        2153   
2  Capital O 44354 Jubilee Banjara Suites           1778    32        1213   
3             OYO 35487 Hotel Meena Grand           2090    32        1425   
4               OYO 16018 Rainbow Saphire           2607    32        1777   

  Distance Rating Number of Rating    Summary       Location       City  
0      1.7    4.4     201 Ratings)  Very Good  Banjara Hills  Hyderabad  
1      1.6    4.4     879 Ratings)  Very Good  Banjara Hills  Hyderabad  
2      1.1    4.4    1258 Ratings)  Very Good  Banjara Hills  Hyderabad  
3      4.0      4     575 Ratings)  Very Good  Banjara Hills  Hyderabad  
4      4.6    4.7     978 Ratings)  Excellent  Banjara Hills  Hyderabad

In [35]:
df['Number of Rating']=df['Number of Rating'].str[:-8]

In [36]:
df.head(5)

Name Original_price Offer Final_price  \
0       OYO 22085 Raj Banjara White House           2459    32        1677   
1         Collection O 42514 Venkat Nagar           3158    32        2153   
2  Capital O 44354 Jubilee Banjara Suites           1778    32        1213   
3             OYO 35487 Hotel Meena Grand           2090    32        1425   
4               OYO 16018 Rainbow Saphire           2607    32        1777   

  Distance Rating Number of Rating    Summary       Location       City  
0      1.7    4.4             201   Very Good  Banjara Hills  Hyderabad  
1      1.6    4.4             879   Very Good  Banjara Hills  Hyderabad  
2      1.1    4.4            1258   Very Good  Banjara Hills  Hyderabad  
3      4.0      4             575   Very Good  Banjara Hills  Hyderabad  
4      4.6    4.7             978   Excellent  Banjara Hills  Hyderabad

In [37]:
df.rename(columns={'Number of Rating':'Number_of_Rating'},inplace=True)

In [38]:
df.dtypes

Name                object
Original_price      object
Offer               object
Final_price         object
Distance            object
Rating              object
Number_of_Rating    object
Summary             object
Location            object
City                object
dtype: object

In [39]:
df['Original_price']=df['Original_price'].astype(float)
df['Offer']=df['Offer'].astype(float)
df['Final_price']=df['Final_price'].astype(float)
df['Distance']=df['Distance'].astype(float)
df['Rating']=df['Rating'].astype(float)
df['Number_of_Rating']=df['Number_of_Rating'].astype(float)

In [40]:
df.dtypes

Name                 object
Original_price      float64
Offer               float64
Final_price         float64
Distance            float64
Rating              float64
Number_of_Rating    float64
Summary              object
Location             object
City                 object
dtype: object

In [41]:
df.drop_duplicates(inplace=True)

In [42]:
df.shape

(11982, 10)

In [43]:
df.set_index('Name')

Original_price  Offer  \
Name                                                                       
OYO 22085 Raj Banjara White House                          2459.0   32.0   
Collection O 42514 Venkat Nagar                            3158.0   32.0   
Capital O 44354 Jubilee Banjara Suites                     1778.0   32.0   
OYO 35487 Hotel Meena Grand                                2090.0   32.0   
OYO 16018 Rainbow Saphire                                  2607.0   32.0   
Collection O 46198 New Sree Krishna                        3964.0   32.0   
OYO 14224 Happy Guest house                                2373.0   32.0   
OYO 62547 Dream Connect Stays                              1941.0   32.0   
OYO 11376 Golden Park                                      2205.0   45.0   
OYO 11316 Athithi Inn                                      2178.0   38.0   
OYO 10272 Hotel Ruma                                       1777.0   32.0   
OYO 18725 Sr Grand                                         1910.0   32.0   
OYO 18302 Pride Banjara                                    2260.0   32.0   
OYO 983 Hotel Surya Residency                              2811.0   32.0   
Capital O 46832 Gaayatri Residency Deluxe                  2098.0   32.0   
OYO 4025 Hotel Meredian Orchid                             1865.0   35.0   
OYO Townhouse 045 Lakdi ka Pul                             5129.0   32.0   
OYO 41751 Hotel Sabharwal                                  1416.0   32.0   
OYO 47332 Hotel Dolphin                                    2509.0   32.0   
OYO 30744 The  Gate                                        2140.0   32.0   
OYO 37906 New Asian Lodge                                  2420.0   32.0   
OYO 26562 Dazzle Stays                                     1730.0   42.0   
SilverKey Executive Stays 27773 IAS Colony                 1639.0   40.0   
OYO 1410 Country Club Begumpet                             4417.0   32.0   
OYO 22907 Hotel Holiday Deluxe                             2407.0   32.0   
OYO 47727 S Grand                                          1427.0   32.0   
Capital O 6422 Luxor Park                                  2737.0   32.0   
OYO 9302 Hotel Baseraa Inn                                 2081.0   32.0   
OYO 8941 Hotel Golden Leaves                               2449.0   32.0   
OYO Flagship 14348 Apollo Hospital                         2746.0   44.0   
...                                                           ...    ...   
OYO 46744 Hotel Yash                                       1481.0   32.0   
OYO 62553 Suvee Palace                                     1294.0   32.0   
OYO 61639 Red Carmine                                      1777.0   32.0   
OYO 44324 Comfort Stay Near Saket Metro                    2989.0   32.0   
OYO Home Luxe 30142 Fabulous Farmstay                      2548.0   49.0   
OYO 22321 Nitin Palace Guest House                         2822.0   32.0   
OYO Townhouse 002 Green Park                               7806.0   32.0   
Capital O 7075 The Oakland Plaza Hotel Deluxe              3078.0   32.0   
OYO 14789 Kempinski Wide                                   3534.0   34.0   
OYO 63837 Ambica Inn                                       2729.0   32.0   
OYO 37512 Spacious Stay Saket Metro                        1981.0   32.0   
OYO 60334 Skp Bed & Breakfast                              2225.0   32.0   
OYO 64271 Singh Hotel 2                                    1140.0   32.0   
OYO 37055 M R Residency Saver                              1481.0   32.0   
Capital O 712 Defence Colony Deluxe                        3738.0   32.0   
OYO Home 47965 Cozy Stay Ignou Road Saket                  2720.0   32.0   
OYO Home 47964 Comfortable  Stay Ignou Road Saket          2720.0   32.0   
OYO 44400 Retreat Inn                                      1878.0   32.0   
OYO Flagship 36485 Smg Heritage Nehru Enclave              1777.0   32.0   
Capital O 39462 Westyard Hotel                             2863.0   32.0   
OYO 40657 Aastha       

In [44]:
df.to_csv('Cleaned_data.csv',index=False,header=True)